internal/services/distribution_account_service.go

```diff
@@ -0,0 +1,114 @@
+package services
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/stellar/go/clients/horizonclient"
+
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/data"
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/services/assets"
+	"github.com/stellar/stellar-disbursement-platform-backend/pkg/schema"
+)
+
+//go:generate mockery --name=DistributionAccountServiceInterface --case=underscore --structname=MockDistributionAccountService --filename=distribution_account_service.go
+type DistributionAccountServiceInterface interface {
+	GetBalances(account *schema.DistributionAccount) (map[data.Asset]float64, error)
+	GetBalance(account *schema.DistributionAccount, asset data.Asset) (float64, error)
+}
+
+type DistributionAccountServiceOptions struct {
+	HorizonClient horizonclient.ClientInterface
+}
+
+type DistributionAccountService struct {
+	strategies map[schema.DistributionAccountType]DistributionAccountServiceInterface
+}
+
+func NewDistributionAccountService(opts DistributionAccountServiceOptions) *DistributionAccountService {
+	stellarNativeDistributionAccSvc := NewStellarNativeDistributionAccountService(opts.HorizonClient)
+
+	strategies := map[schema.DistributionAccountType]DistributionAccountServiceInterface{
+		schema.DistributionAccountTypeEnvStellar:     stellarNativeDistributionAccSvc,
+		schema.DistributionAccountTypeDBVaultStellar: stellarNativeDistributionAccSvc,
+		// TODO [SDP-1232]: schema.DistributionAccountTypeDBVaultCircle: Add Circle distribution account service
+	}
+
+	return &DistributionAccountService{
+		strategies: strategies,
+	}
+}
+
+func (s *DistributionAccountService) GetBalance(account *schema.DistributionAccount, asset data.Asset) (float64, error) {
+	return s.strategies[account.Type].GetBalance(account, asset)
+}
+
+func (s *DistributionAccountService) GetBalances(account *schema.DistributionAccount) (map[data.Asset]float64, error) {
+	return s.strategies[account.Type].GetBalances(account)
+}
+
+var _ DistributionAccountServiceInterface = (*DistributionAccountService)(nil)
+
+type StellarNativeDistributionAccountService struct {
+	horizonClient horizonclient.ClientInterface
+}
+
+var _ DistributionAccountServiceInterface = (*StellarNativeDistributionAccountService)(nil)
+
+func NewStellarNativeDistributionAccountService(horizonClient horizonclient.ClientInterface) *StellarNativeDistributionAccountService {
+	return &StellarNativeDistributionAccountService{
+		horizonClient: horizonClient,
+	}
+}
```

Comment:
What's the advantage of adding this constructor? I usually add constructors. when I want to validate the input variables and return an error when they're not valid.

Do you think a validation would be valuable moving forward, when we have CIrcle and even other strategies?

Reply:
hm I guess not much use for this specific svc implementation, but definitely for the Circle one moving forward